docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+
+```yaml
+apiVersion: core.gardener.cloud/v1beta1
+kind: CloudProfile
+metadata:
+  name: cloudprofile1
+spec:
+  type: <some-provider-name>
+  providerConfig: <some-provider-specific-cloudprofile-config>
+  kubernetes:
+    versions:
+      - version: 1.31.1
+  machineImages:
+    - name: gardenlinux
+      versions:
+        - version: 1443.8.0
+          expirationDate: "2025-02-28T23:59:59Z"
+          minSupportedVersionInPlaceUpdate: 1312.3.0
+        - version: 1443.7.0
+  machineTypes:
+    - name: m5.large
+      cpu: "2"
+      gpu: "0"
+      memory: 8Gi
+      usable: true
+  regions:
+    - name: <region-name>
+```
+
+#### `Shoot` API
+
+A new field `UpdateStrategy` is introduced under `spec.provider.workers[]` in the Shoot spec. This field will be passed on to the worker extension. Once set, changing the `UpdateStrategy` is prohibited through validation. Additionally, when the `InPlaceUpdate/InplaceUpdateOnLabel` update strategy is configured, skipping the worker pool's Kubernetes minor version is not allowed.
+
+```yaml
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+metadata:
+  name: crazy-botany
+  namespace: garden-dev
+spec:
+  secretBindingName: my-provider-account
+  cloudProfile:
+    name: cloudprofile1
+  region: europe-central-1
+  provider:
+    type: <some-provider-name> # {aws,azure,gcp,...}
+    workers:
+      - name: cpu-worker
+        minimum: 5
+        maximum: 5
+        maxSurge: 0
+        maxUnavailable: 2
+        updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel
+        machine:
+          type: m5.large
+          image:
+            name: <some-image-name>
+            version: <some-image-version>
+            providerConfig: <some-machine-image-specific-configuration>
+          architecture: <some-cpu-architecture>
+  kubernetes:
+    version: 1.27.3
+  networking:
+    type: <some-network-extension-name> # {calico,cilium}
+```
+
+#### `Worker` API
+
+A new field `updateStrategy` is introduced under `.spec.pools[]` in the Worker API. This field will be used to set the `machineDeploymentStrategy` for the `machinev1alpha1.MachineDeployment` resource.
+
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Worker
+metadata:
+  name: worker
+  namespace: shoot--foobar--aws
+spec:
+  type: aws
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foobar--aws
+  infrastructureProviderStatus: <infra-status>
+  pools:
+    - name: cpu-worker-1
+      machineType: m4.large
+      machineImage:
+        name: gardenlinux
+        version: 1.2.2
+      minimum: 5
+      maximum: 5
+      maxSurge: 0
+      maxUnavailable: 2
+      updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel
+```
+
+#### `OperatingSystemConfig` API
+
+A new field `version` is introduced in the `OperatingSystemConfig` spec. It specifies the machine image/OS version of the worker pool.
```

Comment:
Is the `version` the desired version? How will this look like during the update? Will an updated node reference a new `version` while a non-updated one will refer to an operating system config with another `version`?

Reply:
It is the desired version. The OSC will have it at all times, and after a successful update, it will be persisted as last applied OSC in the node. The node is always referring to the same OSC. The difference is calculated between the current OSC and the last applied OSC present on the node.